Add Boost Calibration card to V3 Boost webpage

kdav108 opened the pull request:
## Description

This PR adds the UI for the boost calibration card on the `boost` webpage for the `v3 dashboard`. The boost calibration card would allow users to view the distance travelled and the calibrated distance travelled by the bike and also set the calibration value. This PR uses hardcoded values for the distance and calibrated distance values (this will be updated to use correct values in a future PR). However, the input form to set the calibration value is vertically integrated (i.e it sends the calibrated value inserted on the correct `MQTT` topic).

Note: you'll notice that a lot of the code looks similar and that's because I basically copied over code from other files that achieved similar stuff that I wanted 😂. 
## Screenshots

<img width="1278" alt="Screen Shot 2021-02-07 at 6 31 01 pm" src="https://user-images.githubusercontent.com/63642262/107163221-72956300-69fc-11eb-956b-ac5c87a835d2.png">


## Steps to Test

1) Run `yarn install` inside `client/`
2) Use `yarn start` to launch the dashboard webpage 
3) On a seperate terminal navigate to `server/` and run `yarn start` to start the server
4) Use 2 separate terminals to subscribe to `boost/calibrate` and `boost/calibrate/reset` using localhost, here's. quick copy paste stuff you can use: 
`mosquitto_sub -h localhost -t "boost/calibrate"`
`mosquitto_sub -h localhost -t "boost/calibrate/reset"`
5) Navigate to the `boost` page on the `v3 dashboard` and check that the UI looks good. Use the [Figma design](https://www.figma.com/file/HoZBPfj3QqhmbiDUyxn8Ag/Initial-Design?node-id=0%3A1) to compare. 
6) Use the calibration value input form and the `set` and `reset` buttons to verify that they work correctly by checking the subscribed mqtt topics.
7) You can also try to break the input form by entering random stuff and see whether that breaks the code or not.

kdav108 committed:
Add initial UI design

kdav108 committed:
merge master to update new file hierarchy

kdav108 committed:
Refactor BoostCalibrationn function

kdav108 committed:
Add initial simple story

kdav108 committed:
Reformat UI

kdav108 committed:
Add documentation and error handling code for user entered calibration value

kdav108 committed:
Remove actions and simply story

kdav108 committed:
Add Boost Calibration card to the Boost page

kdav108 committed:
Remove unecessary import

kdav108 committed:
Resolve merge conflict

kdav108 committed:
Remove text styling

kdav108 committed:
Add Boost calibration card to the Boost page

kdav108 committed:
Move BoostCalibration script and story to the boost folder inside common

kdav108 committed:
Move powerModel.ts from v2 folder to common

kdav108 committed:
Fix imports and remove BoostCalibrationProps

kdav108 committed:
Comment out code for retrieving distance travelled

kdav108 committed:
Fix import error

kdav108 committed:
Update import statements to reflect shuffle of powerModel.ts

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+            <Card.Body>
+                <Card.Title>Calibration</Card.Title>
+                <Card.Text>
+                    BOOST may use a distance different to the bike&apos;s travelled distance for the purposes of generating power plan data.
```
To avoid the awkward apostrophe thing, we can do a sneaky thing in React. Rather than putting the text as plain html which doesn't like fancy symbols, you can give it a JS string instead: `<Card.Text>{"I'm a JavaScript string"}</Card.Text>`

kdav108 replied:
Wow, that's actually so smart 😂.

kdav108 replied:
when I tried implementing it, `yarn` complains about unnecessary use of curly brackets

hallgchris replied:
That's rather annoying... Happy for you to leave it as is then

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+                    BOOST may use a distance different to the bike&apos;s travelled distance for the purposes of generating power plan data.
+                </Card.Text>
+                <div className="pb-3">
+                    <b>Distance travelled (m) </b> <span className="float-right pr-4" > 30 </span>
```
Can you display the unit after the quantity, rather than before?

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+                    BOOST may use a distance different to the bike&apos;s travelled distance for the purposes of generating power plan data.
+                </Card.Text>
+                <div className="pb-3">
+                    <b>Distance travelled (m) </b> <span className="float-right pr-4" > 30 </span>
```
Can you pass the distance travelled/calibrated distance as a prop, rather than hardcode it here? You can of course still hardcode it in the View and the story. Let me know if you're unsure how to do this

kdav108 replied:
I think I have a general idea of how to go about it, but I was planning to do that in the next PR where I actually try to get the correct value. If you still want it done in this PR, let me know and I can attempt it.

hallgchris replied:
I'd rather it were done in this PR, because it allows us to see how this component will be used and verify that everything still displays properly when given different inputs, and that callbacks are called correctly.

kdav108 replied:
Do we specifically want it passed in from `BoostView` otherwise if the aim is to just use callbacks, I can pass it from `BoostCalibration` if I decide to keep the two classes separate.

hallgchris commented on client/src/views/common/BoostView.tsx:
```diff
@@ -10,7 +11,9 @@ import ContentPage from 'components/common/ContentPage';
 export default function BoostView() {
     return (
         <ContentPage>
```
Add `title="BOOST Configuration"` here for a nice title!

hallgchris commented on client/src/views/common/BoostView.tsx:
```diff
@@ -10,7 +11,9 @@ import ContentPage from 'components/common/ContentPage';
     return (
         <ContentPage>
-            <h1>Hi I&apos;m the Boost View</h1>
+            <div>
```
This `div` is unnecessary

kdav108 replied:
hahaha true, I don't know what I was thinking

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+export default function BoostCalibration() {
+    // const GetDistTravelled = () => {
+    //     return useSensorData("reed_distance").data;
+    // };
```
Remove unused code - this will be passed in via a prop anyway

kdav108 replied:
Should it passed down or I was thinking maybe `BoostCalibration` can just read it off? If it gets passed down, it would have to be passed down by `BoostView` right?

hallgchris replied:
Correct, you'll need to pass it down. The data will be fetched and provided by either `BoostView` or some intermediate container component (I can't remember which we ended up going with elsewhere).

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+            onKeyPress={handleKeyPressed}
+            placeholder="Calibrate distance..."
+            value={calibrationValue}
+        />
```
Have a look at the form validation stuff here: https://react-bootstrap.github.io/components/forms/#forms-validation

Rather than an alert, it would be better to highlight the form red and say something like "Distance must be numerical". You could even do this validation live while the user is typing.

Also consider adding `type="number"` which tells it what you intend to use the form for.

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+function BoostCalibrationInput() {
+    const [calibrationValue, setCalib] = useState('');
+
+    const handleCalibrationChange = useCallback(
```
We seem to use `useCallback` a lot around here but I don't believe it's necessary. (anyone correct me if you believe otherwise...) These functions can (I think) just be normal callbacks.

haotongwang replied:
I think the difference is quite minimal but since it is being used in the form that is re-rendered (I think) when used, having `useCallback` makes it a bit more efficient.

kdav108 replied:
I tried removing the `useCallback` but it generates a warning from `yarn`, so at the moment I'm keeping them as is, but let me know if it's possible to remove those warnings and whether we want to avoid using `useCallbacks` where it's not necessary

hallgchris replied:
On second thoughts I think Nick is right. Don't worry about this :)

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+            alert("Please enter an integer value only!");
+          }
+          else {
+            console.log(calibValue);
```
`console.log` is great for debugging but ideally we won't have any in production!

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+          }
+          else {
+            console.log(calibValue);
+            sendCalibrationValue(parseInt(calibrationValue, 10));
```
Rather than calling the API, for now instead call some callback that is passed in as a prop. This way it's very easy to test if the code is working from within storybooks.

kdav108 replied:
Actually this is something I was confused about. I kind of felt just publishing to dashboard was quite easy to test the code rather than using storybook for which you have to write more code and use callbacks. I think I remember you guys discussing why we went with storybook in one of our weekly meetings, but I can't recall. Was there any specific benefits?

hallgchris replied:
Good question. Writing our components like this allows us to completely detach the UI from the logic, meaning that both can be tested independently. This means we can test the components we write without needing to write any logic beforehand. In this case, some of the power model stuff happens to already be written, but were that not the case, testing whether the buttons work would be more difficult. In components with complex logic driving them, it also makes it much easier to test difficult to access states, or to be able to quickly check that your component looks right for a bunch of different inputs if you make a small change.

Although we haven't currently written any front-end unit tests, having the callbacks as props also makes it possible to automatically test the functionality of the component. For example, we could have a unit test that enters "100" as the calibration distance, clicks the "set" button, and then asserts that it got the correct data submitted in the callback function. This wouldn't be possible if the component just sent the data straight back to the server (which wouldn't even be running for a unit test)

kdav108 replied:
Yeah true that makes sense. With the unit testing though we can easily mock the server side and do the checking that way, but I do get your point!

kdav108 replied:
So is the **calibration difference** published by `BOOST` or the actual **calibrated distance**. Your first comment here  made me think that `BOOST` only provides the calibration difference and so atm the function takes this as input and adds the given instance to display the calibrated distance. Is this the right way around?

hallgchris replied:
Sorry for the confusion, I just checked and right now it publishes the calibration difference. I'm going to be modifying that section of boost shortly though (right now it publishes a query string, going to change it to json) and I could change it if needed.

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+            console.log(calibValue);
+            sendCalibrationValue(parseInt(calibrationValue, 10));
+          }
+          setCalib('');
```
Only reset the form if it passed validation!

kdav108 replied:
I thought it would be a good idea to clear it for the user, since the input is wrong anyways

hallgchris replied:
I would disagree that this is useful. My thinking is that if a user tries to submit something invalid, they'd want to see what what they typed to see what was wrong with it. If you think about forms with validation that you fill in on the internet, most would not clear your input if it's invalid unless it's something sensitive (like a password).

I think just as a general rule in UI/UX design, doing something unexpected is usually not a good experience for the user, in this case clearing a form when they only wanted to submit.

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,86 @@
+ * 
+ * @returns {React.Component} Component
+ */
+function BoostCalibrationInput() {
```
I don't mind strongly either way, but you could combine this component with the parent one. The code might be simpler once you start passing props around.

kdav108 replied:
Yeah that's something I had thought of, I just wasn't sure if I should go for more readable and modular code by having them separate or an easier function instead. From what I've thought so far, the only thing the parent `BoostCalibration` needs is the calibration value for which I can simply pass a function in `BoostCalibration` to `BoostCalibrationInput` that allows the parent to access the value. If this is all I'll need I though having more modular code would be better to test and more readable. let me know what you think!

hallgchris replied:
Since you can justify it I'm happy with you keeping it as is! My personal thoughts were that since the child component was only used exactly once, it actually makes it less readable as you loose the context of how the child relates to the parent unless you read the parent component first. As I said though, this is fine too!

kdav108 replied:
hmmm that actually almost convinces me to go with a single class. I feel like that will make it quite big, but maybe I can try it out and see whether it still is readable to not.

hallgchris commented on client/src/components/common/boost/BoostCalibration.stories.tsx:
```diff
@@ -0,0 +1,14 @@
+  };
+
+  const Template = addArgs<any>((props) => (
+    <BoostCalibration {...props} />
```
Once you've got the buttons calling callbacks passed in as props, supply storybook `action`s as the callbacks so that we can see when the callbacks are triggered from storybook. (See `BoostConfigList.stories.tsx` for an example)

hallgchris commented on client/src/components/common/boost/BoostCalibration.stories.tsx:
```diff
@@ -0,0 +1,14 @@
+import BoostCalibration from './BoostCalibration';
+
+export default {
+    title: 'BoostCalibration',
```
Sort this component with the other boost components as discussed on #team-electrical

hallgchris reviewed (changes requested):
Awesome job with your first front-end PR! You're picking this up much faster than I did :sweat_smile:

kdav108 committed:
Use JavaScript string instead of HTML

kdav108 committed:
Place dimention after the measurement

kdav108 committed:
Add type to form and remove console.log

kdav108 committed:
Remove unecessary div tags and add title to the page

kdav108 committed:
Combine BoostCalibration and BoostCalibrationnInput

kdav108 committed:
Disable users from entering non-integer values in the form

kdav108 committed:
Use form innstead of input to take calibration value input

haotongwang commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,84 @@
+        [onSet]
+    );
+
+    return (
+        <Card >
+            <Card.Body>
+                <Card.Title>Calibration</Card.Title>
+                <Card.Text>
+                    BOOST may use a distance different to the bike&apos;s travelled distance for the purposes of generating power plan data.
+                </Card.Text>
+                <div className="pb-3">
+                    <b>Distance travelled </b> <span className="float-right pr-4" > 30 m </span>
+                </div>
+                <div className="pb-3">
+                  <b>Calibrated distance </b> <span className="float-right pr-4"> 40 m </span>
+                </div>
+                <InputGroup className="mb-1" >
+                    <FormControl
+                        onChange={handleCalibrationChange}
+                        onKeyPress={handleKeyPressed}
+                        placeholder="Calibrate distance..."
+                        value={state.calibrationValue}
+                        type="number"
+                        required
+                    />
+                    <InputGroup.Append className="l5">
+                        <Button className="mr-2" variant="outline-secondary" onClick={onSet}>Set</Button>
+                    </InputGroup.Append>
+                    
+                    <Button variant="outline-danger" >Reset</Button>
+                </InputGroup>
+            </Card.Body>
+        </Card>
+    );
+}
```
Are you planning to separate this component into presentational and logical components in the future when stories are added? For reference, those would be any component where we have `Component.tsx` and `ComponentContainer.tsx`.

kdav108 replied:
I had this function split into 2, but combined them later (see Chris' comment above). I've added stories in this PR in `BoostCalibration.stories.tsx` file, are the 'stories' that you're talking about something different? I'm also not sure of component containers, is this something we're planning to add later?

kdav108 committed:
Make BoostView pass in props for BoostCalibration

kdav108 committed:
Add BoostCalibrationProps

kdav108 committed:
Change data type of calibrationValue variable to integer

kdav108 committed:
Place BoostCalibration Story in the right director in storybook

kdav108 committed:
Wrap submit handler inside useCallback

kdav108 committed:
Remove commented out code

kdav108 committed:
Split the state object into two useState hooks

kdav108 committed:
Remove unecessray code

kdav108 committed:
Remove error message on blur and disbale negative inputs

kdav108 committed:
Add prop for distance travelled and add logic for displayinng calibrated distance

kdav108 committed:
Remove debugginng code

kdav108 commented:
New changes 🥳:

- Combined the functions into one `BoostCalibration` function
- Removed the native JS alert with form validation using React 
- Also added code to prevent non-integers and negative integers from being submitted (Tbh I don't even know what I did but the current implementation doesn't even allow users to type any non-integer value into the field, except the letter 'e' - try it out)
- The set and reset buttons as well as the distance travelled information now utilises props (distance travelled is still hardcoded in `BoostView`)
- The calibrated distance section now correctly shows, well.... the calibrated distance
- I also made it so that if a user accidentally enters the wrong input and then doesn't want to enter any input again, then on mouse click outside of the input form, the error message disappears instead of staying there till the right value is 'set'

kdav108 committed:
Update function comment

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,98 @@
+    
+    const [calibValue, setCalibValue] = useState(0);
+    // Holds the difference between the current distance and the calibrated distance
+    const [calibDiff, setCalibDiff] = useState(0);
```
The calibration difference is actually published by BOOST over MQTT: https://www.notion.so/V3-MQTT-Topics-66e6715d0e1941ffaa82020e5868fbae#28cede6e40a540eabe0e04f2876cbce7

So, I reckon rather than keeping track of it a second time here, take the calibration difference as a prop so that this page can display exactly what BOOST thinks the calibrated distance is, rather than a potentially different value.

kdav108 replied:
Wait so Boost publishes both the normal distance and the calibrated distance?

hallgchris replied:
Well the rear WM publishes the "normal" distance (the distance the bike has travelled), and BOOST publishes the calibrated distance which is just that same distance travelled plus a constant. This component is for telling BOOST what that constant should be.

kdav108 replied:
Right, I should have actually known this, I've looked into the relevant files that do this 😂

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,98 @@
+    const [calibDiff, setCalibDiff] = useState(0);
+
+    const handleCalibrationChange = useCallback(
+        (event) => {
```
Here `event` has `any` type which typescript seems to be okay about but isn't ideal. Can you add types for `event` here and the next few instances? The types are
`React.ChangeEvent<HTMLInputElement>`
`React.FormEvent<HTMLInputElement>`
`React.KeyboardEvent<HTMLInputElement>`

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,98 @@
+    onSet, onReset, distTravelled
+}: BoostCalibrationProps) {
+    // Controls the feedback form
+    const [validated, setValidation] = useState(false);
```
`setValidated` would be a better name, just to keep it similar to the previous variable.

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,98 @@
+        event.preventDefault();
+        setValidation(true);
+    },
+    [calibValue, onSet, setValidation]);
```
I'm getting the following linting warning:
`React Hook useCallback has a missing dependency: 'distTravelled'. Either include it or remove the dependency array.`
Should be an easy fix :)

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -0,0 +1,98 @@
+                            placeholder="Calibrate distance..." 
+                            onKeyPress={handleKeyPressed}
+                            onChange={handleCalibrationChange}
+                            min = "0"
```
I actually think we shouldn't require a positive value. Imagine the bike hasn't started moving yet, but we know the bike will be starting 100m behind where we thought it would when making the power plan. In this case, we'd want to calibrate BOOST to -100m.

hallgchris reviewed (changes requested):
This is coming along nicely, great work! A few small suggestions.

Also, can you format your code with `prettier`? If you're using VSCode I think all you need to do is download the extension, and Ctrl-Shift-P -> format document. (I like to have the format on save setting on too)

kdav108 committed:
Change variable name

kdav108 committed:
Display calibrated distance from prop instead of internally storing it

kdav108 committed:
Fix bug with calibrated distance

kdav108 committed:
Add types for event handlers

kdav108 committed:
Rename variable

kdav108 committed:
Apply prettier formatting

kdav108 committed:
Remove constraint to only input positive calibration distance

hallgchris commented on client/src/components/common/boost/BoostCalibration.tsx:
```diff
@@ -1,98 +1,112 @@
+          distance for the purposes of generating power plan data.
+        </Card.Text>
+        <div className="pb-3">
+          <b>Distance travelled </b>{' '}
+          <span className="float-right pr-4"> {distTravelled} m </span>
+        </div>
+        <div className="pb-3">
+          <b>Calibrated distance </b>{' '}
+          <span className="float-right pr-4">
+            {' '}
+            {distTravelled + calibrationDiff} m{' '}
```
Nit: formatting the code maintained a bunch of spaces you had in the original, but it's made them explicit and a little ugly in the code. My guess is it'll look identical without the spaces. (I'm talking about the `{' '}` instances)

kdav108 replied:
Yeah it's quite annoying, the only reason why I hadn't removed them earlier was because whenever I reformat it's always going to bring them back unfortunately.

hallgchris replied:
That shouldn't be the case, they shouldn't appear again when deleted... Let me know if this keeps happening, interested to know why that would be the case

hallgchris reviewed (changes requested):
Great work, just one tiny nit and we're good!

kdav108 committed:
Remove spaces created from prettier

hallgchris reviewed (approved):
Looks great! Congrats on your first React PR :tada:
